oss-internship-2020/openjpeg/examples/decompress_example.cc

```diff
@@ -0,0 +1,156 @@
+            __NR_lseek,
+        })
+        .AddFile(in_file_)
+        .BuildOrDie();
+  }
+
+ private:
+  std::string in_file_;
+};
+
+int main(int argc, char* argv[]) {
+  gflags::ParseCommandLineFlags(&argc, &argv, true);
+  google::InitGoogleLogging(argv[0]);
+
+  if (argc != 3) {
+    std::cerr << "Usage: " << basename(argv[0]) << " absolute/path/to/INPUT.jp2"
+              << " absolute/path/to/OUTPUT.pnm\n";
+    return EXIT_FAILURE;
+  }
+
+  std::string in_file(argv[1]);
+
+  // Initialize sandbox.
+  Openjp2SapiSandbox sandbox(in_file);
+  absl::Status status = sandbox.Init();
+  CHECK(status.ok()) << "Sandbox initialization failed " << status;
+
+  Openjp2Api api(&sandbox);
+  sapi::v::ConstCStr in_file_v(in_file.c_str());
+
+  // Initialize library's main data-holders.
+  sapi::StatusOr<opj_stream_t*> stream =
+      api.opj_stream_create_default_file_stream(in_file_v.PtrBefore(), 1);
+  CHECK(stream.ok()) << "Stream initialization failed: " << stream.status();
+  sapi::v::RemotePtr stream_pointer(stream.value());
+
+  sapi::StatusOr<opj_codec_t*> codec = api.opj_create_decompress(OPJ_CODEC_JP2);
+  CHECK(codec.ok()) << "Codec initialization failed: " << stream.status();
+  sapi::v::RemotePtr codec_pointer(codec.value());
+
+  sapi::v::Struct<opj_dparameters_t> parameters;
+  status = api.opj_set_default_decoder_parameters(parameters.PtrBoth());
+  CHECK(status.ok()) << "Parameters initialization failed " << status;
+
+  sapi::StatusOr<OPJ_BOOL> bool_status =
+      api.opj_setup_decoder(&codec_pointer, parameters.PtrBefore());
+  CHECK(bool_status.ok() && bool_status.value()) << "Decoder setup failed";
+
+  // Start reading image from the input file.
+  sapi::v::GenericPtr image_pointer;
+  bool_status = api.opj_read_header(&stream_pointer, &codec_pointer,
+                                    image_pointer.PtrAfter());
+  CHECK(bool_status.ok() && bool_status.value())
+      << "Reading image header failed";
+
+  sapi::v::Struct<opj_image_t> image;
+  image.SetRemote(reinterpret_cast<void*>(image_pointer.GetValue()));
+  CHECK(sandbox.TransferFromSandboxee(&image).ok())
+      << "Transfer from sandboxee failed";
+
+  bool_status =
+      api.opj_decode(&codec_pointer, &stream_pointer, image.PtrAfter());
+  CHECK(bool_status.ok() && bool_status.value()) << "Decoding failed";
+
+  bool_status = api.opj_end_decompress(&codec_pointer, &stream_pointer);
+  CHECK(bool_status.ok() && bool_status.value()) << "Ending decompress failed";
+
+  int components = image.data().numcomps;
+
+  // Transfer the read data to the main process.
+  sapi::v::Array<opj_image_comp_t> image_components(components);
+  image_components.SetRemote(image.data().comps);
+  CHECK(sandbox.TransferFromSandboxee(&image_components).ok())
+      << "Transfer from sandboxee failed";
+
+  image.mutable_data()->comps = (opj_image_comp_t*)image_components.GetLocal();
+
+  unsigned int width = reinterpret_cast<unsigned int>(image.data().comps[0].w);
```

Comment:
Brace init or static_cast
For all casting see: https://google.github.io/styleguide/cppguide.html#Casting